File: app/lib/src/config/supabase_public_config.dart
/// Public Supabase client settings from `--dart-define`.
///
/// Both [url] and [publishableKey] must be non-empty. A secret or
/// service-role key must never be supplied here.
class SupabasePublicConfig {
  const SupabasePublicConfig({required this.url, required this.publishableKey});

  static const fromEnvironment = SupabasePublicConfig(
    url: String.fromEnvironment('SUPABASE_URL'),
    publishableKey: String.fromEnvironment('SUPABASE_PUBLISHABLE_KEY'),
  );

  final String url;
  final String publishableKey;

  String get trimmedUrl => url.trim();

  String get trimmedPublishableKey => publishableKey.trim();

  bool get isComplete =>
      trimmedUrl.isNotEmpty && trimmedPublishableKey.isNotEmpty;
}

File: admin/src/config/supabasePublicConfig.ts
export interface SupabasePublicConfig {
  readonly url: string
  readonly publishableKey: string
}

export function trimmedUrl(config: SupabasePublicConfig): string {
  return config.url.trim()
}

export function trimmedPublishableKey(config: SupabasePublicConfig): string {
  return config.publishableKey.trim()
}

/** Both public values must be non-empty. Never pass a service-role key. */
export function isSupabaseConfigComplete(config: SupabasePublicConfig): boolean {
  return trimmedUrl(config) !== '' && trimmedPublishableKey(config) !== ''
}

function envString(value: unknown): string {
  return typeof value === 'string' ? value : ''
}

export function readSupabasePublicConfig(
  env: {
    readonly VITE_SUPABASE_URL?: unknown
    readonly VITE_SUPABASE_PUBLISHABLE_KEY?: unknown
  } = import.meta.env,
): SupabasePublicConfig {
  return {
    url: envString(env.VITE_SUPABASE_URL),
    publishableKey: envString(env.VITE_SUPABASE_PUBLISHABLE_KEY),
  }
}
